volumina/tiling/cache.py

```diff
@@ -0,0 +1,299 @@
+
+    def __getitem__(self, key):
+        return self._caches[key]
+
+    def __iter__(self):
+        return iter(self._caches)
+
+    def __len__(self):
+        return len(self._caches)
+
+    def touch(self, uid):
+        self._caches.move_to_end(uid)
+
+    @property
+    def maxsize(self):
+        return self._policy.size
+
+    def _evict_one(self):
+        self._caches.popitem(last=False)  # removes item in FIFO order
+
+    def _clean(self):
+        while len(self._caches) > self.maxsize:
+            self._evict_one()
+
+
+class TilesCache:
+    """
+    Contains the following caches, with convenience accessor functions for each.
+
+        layerCache: A cache of 'layers', i.e. for every patch a QImage or QGraphicsItem
+                    for every "image source" in the stack
+
+        tileCache: A cache of 'tiles', i.e. the blended QImage objects
+                   that were created by combining all QImage layers from layerCache for a given patch.
+                   (The QGraphicsItem layers do not contribute to the composite tiles in the tileCache.
+                   They are merely stored.)
+
+        layerCacheDirty: A cache of dirty bits for all layers in layerCache
+        layerCacheTimestamp: A cache of timestamps to track how recently each layer was needed.
+
+        tileCacheDirty: A cache of dirty bits for the composite tiles
+                        (i.e. for a given patch, if a single layer in the patch
+                        is dirty, then the tile for that patch is dirty)
+    """
+
+    def __init__(self, first_stack_id, sims, maxstacks):
+        self._lock = threading.Lock()
+        self._sims = sims
+        self._maxstacks = maxstacks
+        self._policy = CachePolicy(maxstacks)
+
+        kwargs = {"first_uid": first_stack_id, "policy": self._policy}
+
+        # [stack_id][tile_id] -> QImage or QGraphicsItem
+        self._tileCache = MultiCache(default_factory=lambda: (None, 0.0), **kwargs)
+
+        # [stack_id][tile_id] -> bool
+        self._tileCacheDirty = MultiCache(default_factory=lambda: True, **kwargs)
+
+        # [stack_id][(ims, tile_id)] -> QImage or QGraphicsItem
+        self._layerCache = MultiCache(**kwargs)
+
+        # [stack_id][(ims, tile_id)] -> bool
+        self._layerCacheDirty = MultiCache(default_factory=lambda: True, **kwargs)
+
+        # [stack_id][(ims, tile_id)] -> float
+        self._layerCacheTimestamp = MultiCache(default_factory=float, **kwargs)
+
+    @property
+    def maxstacks(self):
+        return self._maxstacks
+
+    def set_maxstacks(self, maxstacks):
+        self._policy.set_size(maxstacks)
+
+    def __enter__(self):
+        self._lock.acquire()
+        return self
+
+    def __exit__(self, *args):
+        assert self._lock.locked()
+        self._lock.release()
+
+    def __contains__(self, stack_id):
+        assert self._lock.locked(), "You must claim the _TileCache via a context manager before calling this function."
```

Comment:
I wonder if it makes sense to make such methods that require a lock "synchronized"?

```python
def __contains__(...):
    with self:
        return ...
```

Reply:
I suppose this used as optimization, as RLock for some reason doesn't work with the current codebase it's basically a way to perform multiple cache operations within one "transaction"